action/protocol/staking/handler_stake_migrate.go

```diff
@@ -0,0 +1,181 @@
+	if candidate == nil {
+		return nil, nil, 0, errCandNotExist
+	}
+	actLog, tLog, err := p.withdrawBucket(ctx, staker, bucket, candidate, csm)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	actLogs = append(actLogs, actLog.Build(ctx, nil))
+	transferLogs = append(transferLogs, tLog)
+
+	// call staking contract to stake
+	duration := int64(bucket.StakedDuration / p.getBlockInterval(protocol.MustGetBlockCtx(ctx).BlockHeight))
+	excReceipt, err := p.createNFTBucket(ctx, act, bucket.StakedAmount, big.NewInt(duration), candidate, csm.SM())
+	if err != nil {
+		revertSM()
+		return nil, nil, 0, errors.Wrap(err, "failed to handle execution action")
+	}
+	if excReceipt.Status != uint64(iotextypes.ReceiptStatus_Success) {
+		// TODO: return err or handle error?
+		revertSM()
+		return nil, nil, 0, errors.Errorf("execution failed with status %d", excReceipt.Status)
+	}
+	// add sub-receipts logs
+	actLogs = append(actLogs, excReceipt.Logs()...)
+	transferLogs = append(transferLogs, excReceipt.TransactionLogs()...)
+	return actLogs, transferLogs, excReceipt.GasConsumed, nil
+}
+
+func (p *Protocol) validateStakeMigrate(ctx context.Context, bucket *VoteBucket, csm CandidateStateManager) error {
+	if err := validateBucketOwner(bucket, protocol.MustGetActionCtx(ctx).Caller); err != nil {
+		return err
+	}
+	if !bucket.AutoStake {
+		return &handleError{
+			err:           errors.New("cannot migrate non-auto-staked bucket"),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
+		}
+	}
+	if bucket.isUnstaked() {
+		return &handleError{
+			err:           errors.New("cannot migrate unstaked bucket"),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
+		}
+	}
+	if err := validateBucketWithoutEndorsement(NewEndorsementStateManager(csm.SM()), bucket, protocol.MustGetBlockCtx(ctx).BlockHeight); err != nil {
+		return err
+	}
+	return validateBucketSelfStake(protocol.MustGetFeatureCtx(ctx), csm, bucket, false)
+}
+
+func (p *Protocol) withdrawBucket(ctx context.Context, withdrawer *state.Account, bucket *VoteBucket, cand *Candidate, csm CandidateStateManager) (*receiptLog, *action.TransactionLog, error) {
+	// delete bucket and bucket index
+	if err := csm.delBucketAndIndex(bucket.Owner, bucket.Candidate, bucket.Index); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to delete bucket for candidate %s", bucket.Candidate.String())
+	}
+
+	// update bucket pool
+	if err := csm.CreditBucketPool(bucket.StakedAmount); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to update staking bucket pool %s", err.Error())
+	}
+	// update candidate vote
+	weightedVote := p.calculateVoteWeight(bucket, false)
+	if err := cand.SubVote(weightedVote); err != nil {
+		return nil, nil, &handleError{
+			err:           errors.Wrapf(err, "failed to subtract vote for candidate %s", bucket.Candidate.String()),
+			failureStatus: iotextypes.ReceiptStatus_ErrNotEnoughBalance,
+		}
+	}
+	// clear candidate's self stake if the
```

Comment:
self stake bucket cannot be migrated

Reply:
sure, here is to clear stale status(e.g. endorsement has been expired)